drivers/can/can_mcp251xfd.c

```diff
@@ -0,0 +1,1814 @@
+	uint8_t *reg_byte;
+	uint32_t *reg;
+	int ret;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	reg_byte = mcp251xfd_get_spi_buf_ptr(dev);
+	*reg_byte = 0;
+
+	ret = mcp251xfd_write(dev, MCP251XFD_REG_BYTE_FLTCON(filter_idx), 1);
+	if (ret < 0) {
+		LOG_ERR("Failed to write FLTCON register [%d]", ret);
+		goto done;
+	}
+
+	dev_data->filter_usage &= ~BIT(filter_idx);
+
+	reg = mcp251xfd_get_spi_buf_ptr(dev);
+	reg[0] = 0;
+
+	ret = mcp251xfd_write(dev, MCP251XFD_REG_FLTCON(filter_idx), MCP251XFD_REG_SIZE);
+	if (ret < 0) {
+		LOG_ERR("Failed to write FLTCON register [%d]", ret);
+	}
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+}
+
+static void mcp251xfd_set_state_change_callback(const struct device *dev,
+						can_state_change_callback_t cb, void *user_data)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+
+	dev_data->state_change_cb = cb;
+	dev_data->state_change_cb_data = user_data;
+}
+
+static int mcp251xfd_get_state(const struct device *dev, enum can_state *state,
+			       struct can_bus_err_cnt *err_cnt)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+	uint32_t *reg;
+	int ret = 0;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	reg = mcp251xfd_read_crc(dev, MCP251XFD_REG_TREC, MCP251XFD_REG_SIZE);
+	if (!reg) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	*reg = sys_le32_to_cpu(*reg);
+
+	if (err_cnt != NULL) {
+		err_cnt->tx_err_cnt = FIELD_GET(MCP251XFD_REG_TREC_REC_MASK, *reg);
+		err_cnt->rx_err_cnt = FIELD_GET(MCP251XFD_REG_TREC_TEC_MASK, *reg);
+	}
+
+	if (state == NULL) {
+		goto done;
+	}
+
+	if (!dev_data->started) {
+		*state = CAN_STATE_STOPPED;
+		goto done;
+	}
+
+	if ((*reg & MCP251XFD_REG_TREC_TXBO) != 0) {
+		*state = CAN_STATE_BUS_OFF;
+	} else if ((*reg & MCP251XFD_REG_TREC_TXBP) != 0) {
+		*state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXBP) != 0) {
+		*state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_TXWARN) != 0) {
+		*state = CAN_STATE_ERROR_WARNING;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXWARN) != 0) {
+		*state = CAN_STATE_ERROR_WARNING;
+	} else {
+		*state = CAN_STATE_ERROR_ACTIVE;
+	}
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+	return 0;
+}
+
+static int mcp251xfd_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct mcp251xfd_config *dev_cfg = dev->config;
+
+	*rate = dev_cfg->osc_freq;
+	return 0;
+}
+
+static int mcp251xfd_get_max_filters(const struct device *dev, bool ide)
+{
+	ARG_UNUSED(ide);
+
+	return CONFIG_CAN_MAX_FILTER;
+}
+
+static int mcp251xfd_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct mcp251xfd_config *dev_cfg = dev->config;
+
+	*max_bitrate = dev_cfg->max_bitrate;
+
+	return 0;
+}
+
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+static int mcp251xfd_recover(const struct device *dev, k_timeout_t timeout)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+
+	ARG_UNUSED(timeout);
+
+	if (!dev_data->started) {
+		return -ENETDOWN;
+	}
+
+	return -ENOTSUP;
```

Comment:
Is this due to a hardware limitation?

Reply:
That's right, it only has an auto-recovery mode.